test/fuzz/random.h

```diff
@@ -0,0 +1,84 @@
+#include <random>
+#include <vector>
+
+namespace Envoy {
+namespace Random {
+
+class PsuedoRandomGenerator64 : public RandomGenerator {
+public:
+  PsuedoRandomGenerator64() = default;
+  ~PsuedoRandomGenerator64() override = default;
+
+  void initializeSeed(uint64_t seed) { prng_ = std::make_unique<std::mt19937_64>(seed); }
+
+  // RandomGenerator
+  uint64_t random() override {
+    // Makes sure initializeSeed() was already called
+    ASSERT(prng_ != nullptr);
+    uint64_t to_return = (*prng_.get())();
+    ENVOY_LOG_MISC(trace, "random() returned: {}", to_return);
+    return to_return;
+  }
+  std::string uuid() override { return ""; }
+  std::unique_ptr<std::mt19937_64> prng_;
+};
+
+} // namespace Random
+
+namespace Fuzz {
+class ProperSubsetSelector {
+public:
+  ProperSubsetSelector(const std::string& random_bytestring)
+      : random_bytestring_(random_bytestring) {}
+
+  /**
+   * This function does proper subset selection on a certain number of elements. It returns a vector
+   * of vectors of bytes. Each vector of bytes represents the indexes of a single subset. The
+   * "randomness" of the subset that the class will use is determined by a bytestring passed into
+   * the class. Example: call into function with a vector {3, 5} representing subset sizes, and 15
+   * as number_of_elements. This function would return something such as {{3, 14, 7}, {2, 1, 13, 8,
+   * 6}}
+   */
+
+  std::vector<std::vector<uint8_t>>
+  constructSubsets(const std::vector<uint32_t>& number_of_elements_in_each_subset,
+                   uint32_t number_of_elements) {
+    index_vector_.clear();
+    index_vector_.reserve(number_of_elements);
+    for (uint32_t i = 0; i < number_of_elements; i++) {
+      index_vector_.push_back(i);
+    }
+    std::vector<std::vector<uint8_t>> subsets;
+    for (uint32_t i = 0; i < number_of_elements_in_each_subset.size(); i++) {
+      subsets.push_back(constructSubset(number_of_elements_in_each_subset.at(i)));
+    }
+    return subsets;
+  }
+
+private:
+  // Builds a single subset by pulling indexes off index_vector_
+  std::vector<uint8_t> constructSubset(uint32_t number_of_elements_in_subset) {
+    std::vector<uint8_t> subset;
+    for (uint32_t i = 0; i < number_of_elements_in_subset && index_vector_.size() != 0; i++) {
+      // Index of bytestring will wrap around if it "overflows" past the random bytestring's length.
+      uint64_t index_of_index_vector =
+          random_bytestring_[index_of_random_bytestring_ % random_bytestring_.length()] %
```

Comment:
Won't this always be capped to 256 elements, since it's a `uint8_t` in the string? 

I was originally thinking that we could use a N byte byte string to provide 8*N subset selection by looking at it as a bit vector. I think what you've chosen instead is to make the string a list of indices. Which is fine, it has much better sparse properties. But, I don't think we should cap hosts at 256, this is too small, we want at least O(10k) hosts to model realistic scenarios.

Reply:
Interesting, will look into what number to move this too. @adisuissa @asraa

Reply:
Hey Harvey, I talked about this with Asra in standup and have been thinking about it, I think since this PR has been open a long time that this question should be answered in a subsequent PR. Right now, I'm scared that my fuzzer might have a O(N^2) bottleneck somewhere, but the clear bottleneck is using mock objects, particularly hosts. I think that the way this fuzzer is setup with Mock Hosts, that 4 figure hosts would render this fuzzer useless. I'm actually working on making my Health Checking fuzzer more efficient with the help of flamegraphs, so how about making this a TODO for once I finish making mocks more efficient? I will definitely get to this however, as this is the core functionality of the load balancing fuzzer.

Reply:
Ah, i think this discussion is also orthogonal to the SubsetSelector interface. The SubsetSelector should maybe be able to handle arbitrary sized sets regardless of how many elements the fuzzer wants to set, but a TODO sounds good

Reply:
OK, a TODO is fine, but I think we should make this something that is followed up sooner rather than later, otherwise we will have a false sense of confidence in what we're covering.

Reply:
Hey, once I fix the O(N^2) in my subset selection algorithm I'll play around with this, I'll try to scale this number up within this PR because I would immediately work on this anyway if the PR in it's current state is merged.

Reply:
Getting rid of O(N^2) allowed me to use 10k hosts with no problem :)! However, Asra brings up a good point - my subset selector interface is capped at 256 hosts sort of (uint8_t). I will generalize my random subset selector to be able to return any amount of bits!!!!!! :)

Reply:
Just to document this, I've been messing around with this, and what's currently checked into this PR allows 10k hosts at decent efficiency (of course what's currently checked in has it hardcoded at 256 hosts per priority set). What I'm working on now is figuring out how to handle the fully random bytestring (we wanted entropy across all the bits according to our meeting) in regards to arbitrary index selection (right now 32 bits wide).